File: scripts/import_triaged.py
#!/usr/bin/env python3
"""Import triaged memory candidates into Qdrant.

Reads a triage-annotated review-report.json (produced by triage-findings),
filters for findings with action == "fix", and stores them via embed + upsert.

Usage:
    python3 import_triaged.py report.json [--dry-run]
"""

from __future__ import annotations

import argparse
import hashlib
import json
import sys
from datetime import datetime, timezone
from pathlib import Path
from uuid import uuid4

# Import config from sibling MCP server package
sys.path.insert(
    0,
    str(
        Path(__file__).resolve().parent.parent / "claude-plugin" / "mcp-server" / "src"
    ),
)
from mindojo_mcp.config import QDRANT_COLLECTION  # noqa: E402
from mindojo_mcp.qdrant_utils import embed, ensure_collection, get_qdrant  # noqa: E402

from qdrant_client.models import PointStruct  # noqa: E402


def load_report(path: Path) -> dict:
    """Load and validate the report JSON."""
    data = json.loads(path.read_text())

    if "triage" not in data:
        print("Error: Report has no triage decisions. Run triage-findings first.")
        sys.exit(1)

    return data


def extract_project_from_recommendation(recommendation: str) -> str | None:
    """Parse 'Import as project:<name> memory' → '<name>'."""
    match = __import__("re").search(r"project:(\S+)", recommendation)
    return match.group(1) if match else None


def build_findings_map(report: dict) -> dict[str, dict]:
    """Build finding_id → finding dict from all sections."""
    findings_map: dict[str, dict] = {}
    for section in report.get("findings", []):
        for finding in section.get("findings", []):
            findings_map[finding["id"]] = finding
    return findings_map


def import_memories(report: dict, *, dry_run: bool = False) -> tuple[int, int]:
    """Import fix-marked findings into Qdrant.

    Returns:
        (imported_count, skipped_count)
    """
    if not dry_run:
        ensure_collection(QDRANT_COLLECTION)

    qd = None if dry_run else get_qdrant()

    findings_map = build_findings_map(report)
    triage_decisions = report["triage"].get("decisions", [])

    imported = 0
    skipped = 0

    for decision in triage_decisions:
        finding_id = decision["finding_id"]
        action = decision["action"]

        if action != "fix":
            skipped += 1
            continue

        finding = findings_map.get(finding_id)
        if not finding:
            print(f"Warning: Finding {finding_id} not found in report, skipping")
            skipped += 1
            continue

        # Determine scope from recommendation
        project = extract_project_from_recommendation(finding.get("recommendation", ""))
        user_id = f"project:{project}" if project else "global"

        # Build memory content — title + description
        content = f"{finding['title']}\n\n{finding['description']}"

        metadata = {
            "source_id": finding_id,
            "tags": finding.get("tags", []),
            "imported_from": finding.get("location", ""),
        }

        if dry_run:
            print(f"  [DRY RUN] Would import {finding_id} → {user_id}")
            print(f"    Title: {finding['title']}")
            print(f"    Content length: {len(content)} chars")
        else:
            vectors = embed([content])
            point_id = str(uuid4())
            payload = {
                "data": content,
                "hash": hashlib.md5(content.encode()).hexdigest(),
                "user_id": user_id,
                "created_at": datetime.now(timezone.utc).isoformat(),
                "updated_at": None,
                **metadata,
            }
            qd.upsert(  # type: ignore[union-attr]
                collection_name=QDRANT_COLLECTION,
                points=[PointStruct(id=point_id, vector=vectors[0], payload=payload)],
            )
            print(f"  Imported {finding_id} → {user_id}: {finding['title']}")

        imported += 1

    return imported, skipped


def main() -> None:
    parser = argparse.ArgumentParser(description="Import triaged memories into Qdrant")
    parser.add_argument("report", help="Path to triaged report.json")
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="Show what would be imported without storing",
    )
    args = parser.parse_args()

    report_path = Path(args.report)
    if not report_path.exists():
        print(f"Error: {report_path} not found")
        sys.exit(1)

    report = load_report(report_path)

    print(f"Processing triaged report: {report_path}")
    print(
        f"Triage by: {report['triage'].get('triaged_by', 'unknown')} "
        f"at {report['triage'].get('triaged_at', 'unknown')}"
    )

    imported, skipped = import_memories(report, dry_run=args.dry_run)

    print(f"\nDone: {imported} imported, {skipped} skipped")


if __name__ == "__main__":
    main()

File: rs/mindojo-server/src/import_triaged.rs
//! Import triaged findings into vector store (moved from `mindojo-import-triaged`).

use std::sync::OnceLock;

use chrono::Utc;
use regex::Regex;
use serde::Deserialize;
use tracing::warn;
use uuid::Uuid;

use mindojo_core::config::Settings;
use mindojo_core::embed::FastEmbedProvider;
use mindojo_core::error::{MindojoError, Result as MindojoResult, ResultExt};
use mindojo_core::lancedb_store::LanceDbStore;
use mindojo_core::pipeline::{MEMORIES_TABLE, md5_hex};
use mindojo_core::traits::{EmbeddingProvider, VectorPoint, VectorStore};

use crate::ImportTriagedArgs;

#[derive(Deserialize)]
struct TriageReport {
    findings: Option<Vec<FindingSection>>,
    triage: Option<TriageInfo>,
}

#[derive(Deserialize)]
struct FindingSection {
    #[serde(default)]
    findings: Vec<Finding>,
}

#[derive(Deserialize)]
struct Finding {
    id: String,
    title: String,
    description: String,
    #[serde(default)]
    recommendation: String,
    #[serde(default)]
    tags: Vec<String>,
    #[serde(default)]
    location: String,
}

#[derive(Deserialize)]
struct TriageInfo {
    #[serde(default)]
    triaged_by: String,
    #[serde(default)]
    triaged_at: String,
    #[serde(default)]
    decisions: Vec<TriageDecision>,
}

#[derive(Deserialize)]
struct TriageDecision {
    finding_id: String,
    action: String,
}

fn extract_project_from_recommendation(recommendation: &str) -> Option<String> {
    static RE: OnceLock<Regex> = OnceLock::new();
    let re = RE.get_or_init(|| Regex::new(r"project:(\S+)").unwrap());
    re.captures(recommendation)
        .and_then(|c| c.get(1))
        .map(|m| m.as_str().to_string())
}

pub async fn run(args: &ImportTriagedArgs) -> MindojoResult<()> {
    if !args.report.exists() {
        return Err(MindojoError::Other(format!(
            "Report file not found: {}",
            args.report.display()
        )));
    }

    let raw = std::fs::read_to_string(&args.report)
        .with_context(|| format!("failed to read {}", args.report.display()))?;
    let report: TriageReport = serde_json::from_str(&raw).context("failed to parse report JSON")?;

    let triage = report.triage.as_ref().ok_or_else(|| {
        MindojoError::Other("Report has no triage decisions. Run triage-findings first.".into())
    })?;

    println!("Processing triaged report: {}", args.report.display());
    println!(
        "Triage by: {} at {}",
        if triage.triaged_by.is_empty() {
            "unknown"
        } else {
            &triage.triaged_by
        },
        if triage.triaged_at.is_empty() {
            "unknown"
        } else {
            &triage.triaged_at
        },
    );

    let mut findings_map = std::collections::HashMap::new();
    if let Some(sections) = &report.findings {
        for section in sections {
            for finding in &section.findings {
                findings_map.insert(finding.id.clone(), finding);
            }
        }
    }

    let settings = Settings::load()?;
    settings.ensure_log_dir()?;
    let embedder = FastEmbedProvider::from_settings(&settings)?;

    let store = if !args.dry_run {
        let s = LanceDbStore::open(&settings.lancedb_path).await?;
        s.ensure_table(MEMORIES_TABLE, settings.embed_dims).await?;
        Some(s)
    } else {
        None
    };

    let mut imported = 0u32;
    let mut skipped = 0u32;

    for decision in &triage.decisions {
        if decision.action != "fix" {
            skipped += 1;
            continue;
        }

        let Some(finding) = findings_map.get(&decision.finding_id) else {
            warn!(finding_id = %decision.finding_id, "Finding not found in report, skipping");
            skipped += 1;
            continue;
        };

        let project = extract_project_from_recommendation(&finding.recommendation);
        let user_id = project
            .as_ref()
            .map(|p| format!("project:{}", p))
            .unwrap_or_else(|| "global".to_string());

        let content = format!("{}\n\n{}", finding.title, finding.description);

        if args.dry_run {
            println!("  [DRY RUN] Would import {} -> {}", finding.id, user_id);
            println!("    Title: {}", finding.title);
            println!("    Content length: {} chars", content.len());
        } else {
            let store = store.as_ref().unwrap();
            let vectors = embedder.embed(std::slice::from_ref(&content)).await?;
            let point_id = Uuid::new_v4().to_string();
            let now = Utc::now().to_rfc3339();

            let mut payload = serde_json::Map::new();
            payload.insert("data".into(), serde_json::Value::String(content.clone()));
            payload.insert("hash".into(), serde_json::Value::String(md5_hex(&content)));
            payload.insert("user_id".into(), serde_json::Value::String(user_id.clone()));
            payload.insert("created_at".into(), serde_json::Value::String(now));
            payload.insert("updated_at".into(), serde_json::Value::Null);
            payload.insert(
                "source_id".into(),
                serde_json::Value::String(finding.id.clone()),
            );
            payload.insert(
                "tags".into(),
                serde_json::Value::Array(
                    finding
                        .tags
                        .iter()
                        .map(|t| serde_json::Value::String(t.clone()))
                        .collect(),
                ),
            );
            payload.insert(
                "imported_from".into(),
                serde_json::Value::String(finding.location.clone()),
            );

            let point = VectorPoint {
                id: point_id,
                vector: vectors[0].clone(),
                payload: serde_json::Value::Object(payload),
            };

            store.upsert(MEMORIES_TABLE, &[point]).await?;
            println!(
                "  Imported {} -> {}: {}",
                finding.id, user_id, finding.title
            );
        }

        imported += 1;
    }

    println!("\nDone: {} imported, {} skipped", imported, skipped);
    Ok(())
}
